dags/python_scripts/cr3_extract_diagram_ocr_narrative.py

```diff
@@ -0,0 +1,247 @@
+        sys.stderr.write("Error: Failed to get PDF from the S3 object\n")
+        continue
+
+    # render the pdf into an array of raster images
+    if (args.v):
+        print('Rendering PDF into images');
+    pages = []
+    try:
+        pages = convert_from_bytes(pdf['Body'].read(), 150)
+    except: 
+        sys.stderr.write("Error: PDF Read for crash_id (" + str(crash['crash_id']) + ") failed.\n")
+        continue
+
+
+    # write a rendered image to disk for debugging
+    #pages[1].save('/home/frank/Desktop/page.png')
+
+    if (args.d):
+        if (args.v):
+            print('Excuting a check for a digitally created PDF');
+        digital_end_to_end = True
+        # these pixels are expected to be black on digitally created PDFs
+        pixels = [(110,3520), (3080, 3046), (3050, 2264), (2580, 6056), (1252, 154), (2582, 4166), (1182, 1838)]
+        for pixel in pixels:
+            rgb_pixel = pages[1].getpixel(pixel)
+            if not(rgb_pixel[0] == 0 and rgb_pixel[1] == 0 and rgb_pixel[2] == 0):
+                digital_end_to_end = False
+            if (args.v):
+                print('Pixel' + "(%04d,%04d)" % pixel + ': ' + str(rgb_pixel))
+        if (args.v):
+            print('PDF Digital End to End?: ' + str(digital_end_to_end));
+        if not(digital_end_to_end):
+            if (args.v):
+                sys.stderr.write("Error: Non-digitally created PDF detected.\n")
+            continue
+
+
+    # crop out the narrative and diagram into PIL.Image objects
+    if (args.v):
+        print('Cropping narrative and diagram from images');
+    try:
+        narrative_image = pages[1].crop((96,3683,2580,6049))
+        diagram_image = pages[1].crop((2589,3531,5001,6048))
+    except:
+        sys.stderr.write("Error: Failed to extract the image of the narative and diagram from image in memory\n")
+        continue
+
+
+    # use tesseract to OCR the text
+    if (args.v):
+        print('OCRing narrative')
+    narrative = ''
+    try:
+        narrative = (pytesseract.image_to_string(narrative_image))
+        if (args.v):
+            print("Extracted Text:\n")
+            print(narrative)
+    except:
+        sys.stderr.write("Error: Failed to OCR the narrative\n")
+        continue
+
+
+    # do we want to save a PNG file from the image data that was cropped out where the crash diagram is expected to be?
+    if (args.save_diagram):
+        if (args.v):
+            print('Saving PNG of diagram to S3')
+        try:
+            # never touch the disk; store the image data in a few steps to get to a variable of binary data
+            buffer = io.BytesIO()
+            diagram_image.save(buffer, format='PNG')
+            output_diagram = s3.put_object(Body=buffer.getvalue(), Bucket=args.save_diagram[0], Key=args.save_diagram[1] + '/' + str(crash['crash_id']) + '.png')
+        except:
+            sys.stderr.write("Error: Faild setting s3 object containing the diagram PNG file\n")
+            continue
```

Comment:
Brilliant that it never saves to local disk first! This is gonna help Airflow a lot with space availability. 💾 👍